src/main/java/eu/xenit/alfred/content/gateway/runtime/config/kubernetes/KubernetesApplicationConfigurationRepository.java

```diff
@@ -0,0 +1,77 @@
+
+import eu.xenit.alfred.content.gateway.collections.ObservableMap;
+import eu.xenit.alfred.content.gateway.collections.ObservableMap.MapUpdate;
+import eu.xenit.alfred.content.gateway.runtime.config.ApplicationConfigurationRepository;
+import eu.xenit.alfred.content.gateway.runtime.ApplicationId;
+import eu.xenit.alfred.content.gateway.runtime.config.ComposedApplicationConfiguration;
+import eu.xenit.alfred.content.gateway.runtime.config.ApplicationConfiguration;
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.LabelSelectorBuilder;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+@Slf4j
+@RequiredArgsConstructor
+public class KubernetesApplicationConfigurationRepository implements ApplicationConfigurationRepository {
+
+    private final KubernetesClient client;
+
+    private final String namespace;
+
+    private final ObservableMap<ApplicationId, ComposedApplicationConfiguration> configs = new ObservableMap<>();
+
+    private final SecretMapper<Secret> secretMapper = new Fabric8SecretMapper();
+
+    private static final LabelSelector selector = new LabelSelectorBuilder()
+            .addToMatchLabels("app.kubernetes.io/managed-by", "contentgrid")
+            .addToMatchLabels("app.contentgrid.com/service-type", "gateway")
+            .build();
+
+    @Override
+    public Mono<ApplicationConfiguration> getApplicationConfiguration(String appId) {
+        return Mono.fromSupplier(() -> configs.get(appId));
+    }
+
+    @Override
+    public Flux<MapUpdate<ApplicationId, ApplicationConfiguration>> observe() {
+        return this.configs.observe()
+                .map(update -> new MapUpdate<>(update.getType(), update.getKey(), update.getValue()));
+    }
+
+    public void watchSecrets() {
+        client.secrets()
+                .inNamespace(namespace)
+                .withLabelSelector(selector)
+                .watch(new Watcher<>() {
+                    @Override
+                    public void eventReceived(Action action, Secret resource) {
+                        secretMapper.apply(resource).ifPresent(fragment -> {
+                            var appConfig = configs.computeIfAbsent(fragment.getApplicationId(),
+                                    ComposedApplicationConfiguration::new);
+
+                            log.info("secret {} {} [app-id:{}] # kubectl get secrets {} -o yaml", action, resource.getMetadata().getName(), fragment.getApplicationId(), resource.getMetadata().getName());
+
+                            switch (action) {
+                                case ADDED, MODIFIED -> appConfig.putFragment(fragment);
```

Comment:
Your `ObservableMap` will only emit an update at the point when a `ComposedApplicationConfiguration` is inserted, not when new fragments are added.

You're currently not subscribing to the `ObservableMap` anywhere, but it is something important to take into account, as your subscribers will not receive a notification when the `ApplicationConfiguration` is changed by addition or removal of a fragment

Reply:
Good point - maybe I should make this `ComposedApplicationConfiguration` immutable, so we get proper update-events